src/router.ts

```diff
@@ -119,69 +118,81 @@ export function findRouteGivenExactInput(
   startToken: string,
   endToken: string,
   amountIn: number,
-  maxHops = 3,
+  maxHops: number,
 ): Route | null {
   const tokens = Object.keys(graph);
+  // distances[token][hop] is the maximum amount of token that can be received given hop number
   let distances: Distances = {};
+  // predecessors[token][hop] is the previous hop of the optimal path
   let predecessors: Predecessors = {};
-  let hops: Hops = {};
 
+  const defaultDistance = -Infinity;
   for (const token of tokens) {
-    distances[token] = -Infinity;
-    predecessors[token] = null;
-    hops[token] = 0;
+    distances[token] = {};
+    predecessors[token] = {};
   }
-  distances[startToken] = amountIn;
+  distances[startToken][0] = amountIn;
 
   for (let i = 0; i < maxHops; i++) {
     const newDistances = { ...distances };
-    const newPredecessors = { ...predecessors };
-    const newHops = { ...hops };
 
     for (const [token, edges] of Object.entries(graph)) {
       for (const edge of edges) {
         const fromToken = edge.pool.coinAddresses[edge.fromIndex];
-        // Skip if fromToken is the endToken. This prevents cycles.
-        if (fromToken === endToken) continue;
-
         const toToken = edge.pool.coinAddresses[edge.toIndex];
-        if (distances[fromToken] !== -Infinity) {
+        if (fromToken === endToken || toToken === startToken) continue; // This prevents cycles
+
+        for (const [hopStr, distance] of Object.entries(distances[fromToken])) {
```

Comment:
@wjw12 I think we can avoid this loop. Because the most outside loop `for (let i = 0; i < maxHops; i++)` could be think as calculate the maximum distance for `i+1` hop, with the data of `i` hop. Then it's also no need to copy the `const newDistances = { ...distances };`

Something like this:
```
const newDistance = calcOutGivenIn(
    distances[fromToken][i],
    edge.pool,
    edge.fromIndex,
    edge.toIndex,
);

if(newDistance > distances[fromToken][i+1] || distances[fromToken][i+1] === undefined) {
               ...
}
```

Reply:
I see you're right! standard dynamic programming